File: java/base/src/main/vsdk/toolkit/media/solidTexture/TextureUtils.java
package vsdk.toolkit.media.solidTexture;

import vsdk.toolkit.common.color.ColorRgba;
import vsdk.toolkit.common.linealAlgebra.Vector3Dd;
import vsdk.toolkit.common.statistics.SolidTextureStatistics;
import vsdk.toolkit.media.RGBAColorPalette;
import vsdk.toolkit.media.solidTexture.procedural.ProceduralNoise;

public final class TextureUtils
{
    private static final long WAVE_RANDOM_MASK = 0x7fffl;
    private static final float WAVE_RANDOM_DIVISOR = (float)WAVE_RANDOM_MASK;

    private ProceduralNoise proceduralNoise;
    private double[] frequencyInstance;
    private Vector3Dd[] waveSourcesInstance;

    public TextureUtils()
    {
        proceduralNoise = null;
        frequencyInstance = new double[0];
        waveSourcesInstance = new Vector3Dd[0];
    }

    public TextureUtils(SolidTextureStatistics stats)
    {
        this();
        proceduralNoise = new ProceduralNoise(stats);
    }

    public void initialize(SolidTextureStatistics stats)
    {
        proceduralNoise = new ProceduralNoise(stats);
    }

    public ProceduralNoise getProceduralNoise()
    {
        if ( proceduralNoise == null ) {
            proceduralNoise = new ProceduralNoise();
        }
        return proceduralNoise;
    }

    public static double floorInline(double x)
    {
        return x >= 0.0 ? Math.floor(x) : 0.0 - Math.floor(0.0 - x) - 1.0;
    }

    public static double fabsInline(double x)
    {
        return x < 0.0 ? 0.0 - x : x;
    }

    public double[] waveFrequency()
    {
        return frequencyInstance;
    }

    public Vector3Dd[] waveSources()
    {
        return waveSourcesInstance;
    }

    public static void computeColor(ColorRgba color, RGBAColorPalette colorMap, double value)
    {
        color.set(colorMap.evalLinear(value));
    }

    public void initializeNoise(int numberOfWaves)
    {
        ProceduralNoise noise = getProceduralNoise();
        noise.initialize();
        CRandom random = new CRandom(0L);
        for ( int i = 0; i < 4096; i++ ) {
            random.next();
        }
        frequencyInstance = new double[numberOfWaves];
        waveSourcesInstance = new Vector3Dd[numberOfWaves];

        for ( int i = 0; i < numberOfWaves; i++ ) {
            Vector3Dd point = noise.differentialNoise(i, 0.0, 0.0).normalized();
            waveSourcesInstance[i] = point;
            frequencyInstance[i] = (random.next() & WAVE_RANDOM_MASK) / WAVE_RANDOM_DIVISOR + 0.01;
        }
    }

    private static final class CRandom
    {
        private long state;

        CRandom(long seed)
        {
            state = seed & 0x7fffffffL;
        }

        int next()
        {
            state = (state * 1103515245L + 12345L) & 0x7fffffffL;
            return (int)((state >>> 16) & 0x7fffL);
        }
    }
}

File: typescript/base/src/main/vsdk/toolkit/media/solidTexture/TextureUtils.ts
import { ColorRgba } from "../../common/color/ColorRgba.js";
import { Vector3Dd } from "../../common/linealAlgebra/Vector3Dd.js";
import type { SolidTextureStatistics } from "../../common/statistics/SolidTextureStatistics.js";
import type { RGBAColorPalette } from "../RGBAColorPalette.js";
import { ProceduralNoise } from "./procedural/ProceduralNoise.js";

/**
Java's private nested `CRandom`. `ProceduralNoise` declares its own copy of the
same generator, and so does this file, because the Java sources do; see the
copy there for why the multiplication is split at 2^16.
*/
class _CRandom {
    private static readonly MULTIPLIER = 1103515245;
    private static readonly INCREMENT = 12345;
    private static readonly MODULUS = 0x80000000;

    private state: number;

    public constructor(seed: number) {
        this.state = seed & 0x7fffffff;
    }

    public next(): number {
        const high: number = Math.floor(this.state / 0x10000);
        const low: number = this.state % 0x10000;
        const product: number = ((high * _CRandom.MULTIPLIER) % 0x8000) * 0x10000 + low * _CRandom.MULTIPLIER;
        this.state = (product + _CRandom.INCREMENT) % _CRandom.MODULUS;
        return Math.floor(this.state / 0x10000) & 0x7fff;
    }
}

/**
Port of `vsdk.toolkit.media.solidTexture.TextureUtils`.

The shared services of the solid-texture family: the lazily built
{@link ProceduralNoise}, the `floor` and `abs` helpers the texture fixtures
call by name, the palette lookup, and the wave-source table
{@link BumpTextureFixture} rides on.

Java's `TextureUtils(SolidTextureStatistics)` constructor and its no-argument
one differ only in whether the noise generator is built eagerly with counters
attached; TypeScript spells that as one optional parameter.
*/
export class TextureUtils {
    private static readonly WAVE_RANDOM_MASK = 0x7fff;
    private static readonly WAVE_RANDOM_DIVISOR = TextureUtils.WAVE_RANDOM_MASK;

    private proceduralNoise: ProceduralNoise | null = null;
    private frequencyInstance: Float64Array = new Float64Array(0);
    private waveSourcesInstance: Vector3Dd[] = [];

    public constructor(stats: SolidTextureStatistics | null = null) {
        if (stats !== null) {
            this.proceduralNoise = new ProceduralNoise(stats);
        }
    }

    public initialize(stats: SolidTextureStatistics): void {
        this.proceduralNoise = new ProceduralNoise(stats);
    }

    public getProceduralNoise(): ProceduralNoise {
        if (this.proceduralNoise === null) {
            this.proceduralNoise = new ProceduralNoise();
        }
        return this.proceduralNoise;
    }

    public static floorInline(x: number): number {
        return x >= 0.0 ? Math.floor(x) : 0.0 - Math.floor(0.0 - x) - 1.0;
    }

    public static fabsInline(x: number): number {
        return x < 0.0 ? 0.0 - x : x;
    }

    public waveFrequency(): Float64Array {
        return this.frequencyInstance;
    }

    public waveSources(): Vector3Dd[] {
        return this.waveSourcesInstance;
    }

    public static computeColor(color: ColorRgba, colorMap: RGBAColorPalette, value: number): void {
        color.set(colorMap.evalLinear(value));
    }

    public initializeNoise(numberOfWaves: number): void {
        const noise: ProceduralNoise = this.getProceduralNoise();
        noise.initialize();
        const random = new _CRandom(0);
        for (let i = 0; i < 4096; i++) {
            random.next();
        }
        this.frequencyInstance = new Float64Array(numberOfWaves);
        this.waveSourcesInstance = new Array<Vector3Dd>(numberOfWaves);

        for (let i = 0; i < numberOfWaves; i++) {
            const point: Vector3Dd = noise.differentialNoise(i, 0.0, 0.0).normalized();
            this.waveSourcesInstance[i] = point;
            this.frequencyInstance[i] =
                (random.next() & TextureUtils.WAVE_RANDOM_MASK) / TextureUtils.WAVE_RANDOM_DIVISOR + 0.01;
        }
    }
}
